addons/stock/models/stock_inventory.py

```diff
@@ -319,6 +332,135 @@ def _get_inventory_lines_values(self):
+                                      'location_ids': location})
+            self.create(location_vals)
+        # conflict inventories handled in separate function so inventories can also be updated when user opens inventory menuitem
+        self._run_conflict_inventory_tasks(company_domain=company_domain)
+
+    @api.model
+    def _run_conflict_inventory_tasks(self, **kwargs):
+        """ Updates/creates/deletes conflict inventories. Conflict inventories include:
+                - negative quantities values, created per warehouse
+            Note that an in progress inventory will prevent an update/unlink when its
+            corresponding values are out of date.
+        """
+        company_domain = kwargs.get('company_domain')
+        if not company_domain:
+            company_domain = [('company_id', 'in', self.env.user.company_ids.ids)]
+        # negative quantity check
+        inventory_vals_to_create = []
+        updated_invs = self.env['stock.inventory']
+        existing_conflict_invs = self.search(expression.AND([[('state', 'in', ['draft', 'confirm']),
+                                                              ('is_conflict_inventory', '=', True)],
+                                                             company_domain]))
+        neg_quants = self.env['stock.quant'].search(expression.AND([[('quantity', '<', 0.0),
+                                                                     ('location_id.usage', 'in', ['internal', 'transit']),
+                                                                     ('company_id', 'in', self.env.user.company_ids.ids)],
+                                                                    company_domain]))
+        company_ids = neg_quants.mapped('company_id')
+        warehouse_ids = self.env['stock.warehouse'].search([('company_id', 'in', company_ids.ids)])
+        warehouse_locations = self.env['stock.location'].search([('id', 'child_of', warehouse_ids.mapped('view_location_id').ids)])
+        company_to_warehouses = defaultdict(lambda: self.env['stock.warehouse'])
+        warehouse_to_locations = defaultdict(lambda: self.env['stock.location'])
+        for warehouse in warehouse_ids:
+            company_to_warehouses[warehouse.company_id] |= warehouse
+            warehouse_to_locations[warehouse] = warehouse_locations.filtered(lambda l: any(int(location_id) == warehouse.view_location_id.id for location_id in l.parent_path.split('/')[:-2]))
+        for company_id in company_ids:
+            # separate auto-generated inventories by warehouse
+            for warehouse in company_to_warehouses[company_id]:
+                # avoid conflicting inventories! Wait until the next time this is run after the previous neg qty inventory is completed
+                if existing_conflict_invs.filtered(lambda i: i.state == 'confirm' and i.company_id == company_id and i.location_ids & warehouse_to_locations[warehouse]):
```

Comment:
I'm a little suspicious about this `&`, why not a classic `and` ?

Reply:
We want to avoid creating a new negative qty inventory for a warehouse that already has a confirmed negative qty inventory, so can't we use `&` to check if there's an intersection between any of the existing confirmed negative qty inventories' locations and the warehouse's locations? By classic `and` I assume you mean `i.locations_ids == warehouse_to_locations([warehouse])` which will fail if a top level location is added/removed from that warehouse (but maybe that's OK?).